### DISCUSSION TITLE
Which step would one take to integrate an existing Makefile with moulti?

### DISCUSSION THREAD

**Discussion by MartyLake** [xavierog/moulti#4] [Q&A] [answered]
Hello,
Thanks for creating this software!
Which step would one take to integrate an existing Makefile with moulti?

One possibility might be to use `script scrip.txt make configure build` to get the make output to go to a text fail, then `tail` that file to moulti.
However, how to differentiate Makefile steps ? Should I instrument the makefile to emit markers ?

Also, how to do all that in the same terminal instead of multiple ones?

Best,

**Comment by xavierog** [comment]
> Which step would one take to integrate an existing Makefile with moulti?
> 
> One possibility might be to use `script script.txt make configure build` to get the make output to go to a text file

Do you really need `script` for that? Regular output redirection should do, especially if you can convince your tools to keep outputting ANSI colors.

> then `tail` that file to moulti.

Same: technically, you do not need an intermediate file (but feel free to generate one anyway if that makes sense in your workflow).

> However, how to differentiate Makefile steps ? Should I instrument the makefile to emit markers ?

I see two main approaches:
Approach 1: editing the Makefile so it runs `moulti step add` whenever necessary; but then it becomes necessary to pipe all output to a `moulti pass` command until the next `moulti step add`. I'll have a look at the GNU make documentation but right now, that does not strike me as a simple and straightforward approach.

Approach 2: as you suggest, having markers in the output of `make` makes it easier to write a wrapper. Actually, one of my ideas for the future of Moulti is to provide a wrapper tool that reads a format like that:
```
===== divider1 =====
Lorem ipsum dolor sit amet, consectetur adipiscing elit.
##### title1 #####
# moulti: classes=warning min_height=3 max_height=3 top_text="this format still" bottom_text="needs some work"
Curabitur ultricies dolor augue, id eleifend erat varius ut.
Etiam eget dapibus lectus, in gravida sapien. Aliquam nec diam mi.
Interdum et malesuada fames ac ante ipsum primis in faucibus.
##### title2 #####
# moulti: classes=success
Mauris in pellentesque lorem. Nam id libero at libero malesuada euismod non vel turpis.
Aenean dolor sem, suscipit at risus sed, tempus blandit nisl. Nullam ut tincidunt metus, non interdum urna. 
```
... and turns each title into a Moulti step. That way, Moulti users could "simply" reformat the output of their existing scripts to integrate them with Moulti. This is essentially a less specialized variant of what `moulti diff` does already.

> Also, how to do all that in the same terminal instead of multiple ones?

Although the quickstart makes you play with two terminals at first, the documentation also details `moulti run` in the [Shell scripting](https://github.com/xavierog/moulti/blob/ef9cdf16b57e2803a70017f7df8075cc67f11869/Documentation.md#shell-scripting) section.

**Comment by MartyLake** [comment]
> Do you really need `script` for that? Regular output redirection should do, especially if you can convince your tools to keep outputting ANSI colors.
> Same: technically, you do not need an intermediate file (but feel free to generate one anyway if that makes sense in your workflow).

I would very much like to avoid doing so, I am happy to read that you think it can be avoided altogether.

> I see two main approaches: Approach 1: editing the Makefile so it runs `moulti step add` whenever necessary; but then it becomes necessary to pipe all output to a `moulti pass` command until the next `moulti step add`. I'll have a look at the GNU make documentation but right now, that does not strike me as a simple and straightforward approach.

This is what I undertood I should do reading the Documentation.md, and it seemed unrealistic to create a whole new makefile system to accomodate for moulti.

> Although the quickstart makes you play with two terminals at first, the documentation also details `moulti run` in the [Shell scripting](https://github.com/xavierog/moulti/blob/ef9cdf16b57e2803a70017f7df8075cc67f11869/Documentation.md#shell-scripting) section.

See answer for Approach1.

> Approach 2: as you suggest, having markers in the output of `make` makes it easier to write a wrapper. Actually, one of my ideas for the future of Moulti is to provide a wrapper tool that reads a format like that:
> 
> ```
> ===== divider1 =====
> Lorem ipsum dolor sit amet, consectetur adipiscing elit.
> ##### title1 #####
> # moulti: classes=warning min_height=3 max_height=3 top_text="this format still" bottom_text="needs some work"
> Curabitur ultricies dolor augue, id eleifend erat varius ut.
> Etiam eget dapibus lectus, in gravida sapien. Aliquam nec diam mi.
> Interdum et malesuada fames ac ante ipsum primis in faucibus.
> ##### title2 #####
> # moulti: classes=success
> Mauris in pellentesque lorem. Nam id libero at libero malesuada euismod non vel turpis.
> Aenean dolor sem, suscipit at risus sed, tempus blandit nisl. Nullam ut tincidunt metus, non interdum urna. 
> ```
> 
> ... and turns each title into a Moulti step. That way, Moulti users could "simply" reformat the output of their existing scripts to integrate them with Moulti. This is essentially a less specialized variant of what `moulti diff` does already.

That would be ideal! the moulti instructions could even be hidden behind a "erase current line" terminal command so that it is not visible when running on a system where moulti is not installed.

I had a great experience with the klogg app https://klogg.filimonov.dev/ where you can open any log, then search for things and matched lines would list in a window bellow. You could also define regexes and styles to be applied (fg / bg color, bold etc).

I guess one direction moulti development could take is follow that approach where a user create some configuration file that corresponds to the way its existing scripts already work, and then wrap the launch of such script.

For example, launching make with `-d` flags yields some interesting strings to grep:
```
    Must remake target `build-Debug'.
ninja: Entering directory `output/MyApp/Debug'

[0/20] SOLINK libMyAppLib.dylib, POSTBUILDS[K
[1/20] SOLINK libMyAppLib.dylib, POSTBUILDS[K
    Successfully remade target file `build-Debug'.
   Finished prerequisites of target file `build'.
  Must remake target `build'.
  Successfully remade target file `build'.
 Finished prerequisites of target file `all'.
Must remake target `all'.
Successfully remade target file `all'.
```

Here the steps are started by "Must remake target" and completed by "Successfully remade target", where "[X/Y]" is the progress of such target.

If moulti has some DSL to express that in the command line, or with some ad-hoc config file ,it would be easier to integrate with existing machinery.

Or I could create a filter that calls moulti appropriately when such strings appear, but I am not even sure how to do that.

**Comment by xavierog** [comment, marked as answer]
> Or I could create a filter that calls moulti appropriately when such strings appear, but I am not even sure how to do that.

Like this?

`moultimake.bash`:
```bash
#!/usr/bin/env bash

# Instruct the Moulti instance to harvest any output not explicitly assigned to
# a step:
export MOULTI_RUN_OUTPUT=harvest
export MOULTI_INSTANCE="make"
[ "${MOULTI_RUN}" ] || exec moulti run -- "$0" "$@"

function process_lines {
	local step_counter=0
	while read -r line; do
		if [[ "${line}" =~ 'Must remake target' ]]; then
			# New target: create a new Moulti step:
			local title=$(sed 's/Must remake target `\(.*\)'\''./\1/' <<< "${line}")
			((++step_counter))
			local step_id="make_$$_${step_counter}"
			moulti step add "${step_id}" --title="${title}" --bottom-text=' ' < /dev/null
			# Redirect stdout to `moulti pass our_new_step`
			exec > >(moulti pass "${step_id}")
		fi
		[[ "${line}" =~ 'Successfully remade target' ]] && moulti step update "${step_id}" --classes=success
		[[ "${line}" =~ error ]] && moulti step update "${step_id}" --classes=error
		printf '%s\n' "${line}"
	done
}

# Instance title:
moulti set --title="$*"

# Run the command given on the command-line and process each line of its
# output:
"$@" 2>&1 | process_lines
```

Usage:
```console
chmod +x moultimake.bash
./moultimake.bash make -d
```

Regarding the rest of your suggestions:  thank you, I will take all of this into account when designing the next features.
The thing with Moulti is: [there will always be more that one way to do it](https://en.wiktionary.org/wiki/TMTOWTDI). Right now, Moulti requires writing a controlling script like the one above, except for a couple of uses (like Ansible playbooks and unified diff) where it already provides wrappers out of the box. As time goes on, Moulti should provide more of such wrappers, including some matching your use cases.

**Comment by MartyLake** [comment]
It works!!! That’s so cool! I think that example that you just created can easily be adapted to other "filter" use cases, and should be put in the documentation somewhere close to the beginning. :)

**Comment by xavierog** [comment]
>  It works!!! That’s so cool!

Glad you like it.

> I think that example that you just created can easily be adapted to other "filter" use cases, and should be put in the documentation somewhere close to the beginning. :)

It will definitely serve as a basis for my next reflections. Specifically:
- I can turn it into some kind of mini-framework for users who prefer *imperative* approaches.
- it is a good tool to explore what a *declarative* approach should cover -- e.g. you can already notice a number of arbitrary decisions I took when writing this script: a good declarative format should provide enough flexibility to account for most, if not all use cases.

**Comment by MartyLake** [comment]
> Glad you like it.

Yes! And it was easy now with this template to complete with some progress bar \o/

see `grep` and `sed` lines:
```bash
#!/usr/bin/env bash

# Instruct the Moulti instance to harvest any output not explicitly assigned to
# a step:
export MOULTI_RUN_OUTPUT=harvest
export MOULTI_INSTANCE="make"
[ "${MOULTI_RUN}" ] || exec moulti run -- "$0" "$@"

function process_lines {
	local step_counter=0
	while read -r line; do
		if [[ "${line}" =~ 'Must remake target' ]]; then
			# New target: create a new Moulti step:
			local title=$(sed 's/Must remake target `\(.*\)'\''./\1/' <<<"${line}")
			((++step_counter))
			local step_id="make_$$_${step_counter}"
			moulti step add "${step_id}" --title="${title}" --bottom-text=' ' </dev/null
			# Redirect stdout to `moulti pass our_new_step`
			exec > >(moulti pass "${step_id}")
		fi
		if echo ${line} | grep -q '^\[[0-9]*\/[0-9]*\]'; then
			local progress_status=$(echo ${line} | sed -e 's/^\[\([0-9]*\)\/\([0-9]*\).*/\1/')
			local progress_target=$(echo ${line} | sed -e 's/^\[\([0-9]*\)\/\([0-9]*\).*/\2/')
			moulti set --progress-bar --progress="${progress_status}" --progress-target="${progress_target}"
		fi
		[[ "${line}" =~ 'Successfully remade target' ]] && moulti step update "${step_id}" --classes=success && moulti set --no-progress-bar
		[[ "${line}" =~ error ]] && moulti step update "${step_id}" --classes=error
		printf '%s\n' "${line}"
	done
}

# Instance title:
moulti set --title="$*"

# Run the command given on the command-line and process each line of its
# output:
"$@" 2>&1 | process_lines
```

Right now, when some steps have happened, I cannot know what happens or happened until I scroll with the mouse over moulti.

klogg has a `follow` mode where the window scrolls to the latest line. that mode is toggled by pressing `f`.

While moulti already scrolls inside steps, could moulti also have something similar to scroll finished/inactive steps ? (should I open new issues?)


> It will definitely serve as a basis for my next reflections. Specifically:
> 
>     * I can turn it into some kind of mini-framework for users who prefer _imperative_ approaches.

That would be my case I guess.

>     * it is a good tool to explore what a _declarative_ approach should cover -- e.g. you can already notice a number of arbitrary decisions I took when writing this script: a good declarative format should provide enough flexibility to account for most, if not all use cases.

When I see how concise this moultimake.bash is, I wonder if it is necessary at all. Do you have any idea how a declarative file would look like?

**Comment by xavierog** [comment]
> Right now, when some steps have happened, I cannot know what happens or happened until I scroll with the mouse over moulti.
> 
> klogg has a `follow` mode where the window scrolls to the latest line. that mode is toggled by pressing `f`.
> 
> While moulti already scrolls inside steps, could moulti also have something similar to scroll finished/inactive steps ? (should I open new issues?)

Moulti has programmatic scrolling too. Refer to the documentation for details but TL;DR:
- add `--scroll-on-activity=-1` to the `moulti step add` command
- hit 'l' to suspend programmatic scrolling (the main scrollbar turns green)
- hit 'l' again to restore it (the main scrollbar turns grey again)

**Comment by xavierog** [comment]
> When I see how concise this moultimake.bash is, I wonder if it is necessary at all.

That will be part of my reflections too.

> Do you have any idea how a declarative file would look like?

The default approach would be a serializable Python dict written as JSON/YAML/TOML that describes:
- named patterns (e.g. `section_title: '^#{10} (?<title>.+) #{10}$'`)
- pattern groups (typically optional, purely for the sake of convenience)
- actions (e.g. new step with these parameters, new divider with those options, set instance title or progress bar status)
- action groups (typically optional, purely for the sake of convenience)
- pattern/action associations (e.g. "upon reading a pattern from this pattern group, run those action groups")
- what time to trigger the coffee machine

\+ support for using regex captures and environment variables in actions.

I could implement that by myself OR I could look for a tool/library/framework that already provides this kind of logic.

**Comment by MartyLake** [comment]
> Moulti has programmatic scrolling too. Refer to the documentation for details but TL;DR:
>     * add `--scroll-on-activity=-1` to the `moulti step add` command

That’s very neat! I wonder how I missed that when searching in the documentation. Thanks!

>     * hit 'l' to suspend programmatic scrolling (the main scrollbar turns green)
>     * hit 'l' again to restore it (the main scrollbar turns grey again)

I tried those, but I have troubles understanding if scroll stops, or if scripts takes time. I will use `--scroll-on-activity` for now anyway so I will figure this later.

> The default approach would be a serializable Python dict written as JSON/YAML/TOML that describes:
> 
>     * named patterns (e.g. `section_title: '^#{10} (?<title>.+) #{10}$'`)
>     * pattern groups (typically optional, purely for the sake of convenience)
>     * actions (e.g. new step with these parameters, new divider with those options, set instance title or progress bar status)
>     * action groups (typically optional, purely for the sake of convenience)
>     * pattern/action associations (e.g. "upon reading a pattern from this pattern group, run those action groups")
>     * what time to trigger the coffee machine
> 
> + support for using regex captures and environment variables in actions.
> 
> I could implement that by myself OR I could look for a tool/library/framework that already provides this kind of logic.

Reading all of this, I wonder if a python API might be easier to implement (and more powerful) than JSON/YAML/TMOL.

This api would read like declarative, but really just call python functions, a bit like the [cmake syntax](https://cmake.org/cmake/help/latest/command/target_include_directories.html#command:target_include_directories).

```python3
import moulti as m

m.detector.add_new_step_rule("Must remake target `\(.*\)'\'", auto_increment_id=True, show_progress_bar=False)
m.detector.add_step_success_rule(".*Successfully remade target.*")
m.detector.add_step_error_rule(".*error.*")
m.detector.add_step_status_progress_rule("^\[\([0-9]*\)\/[0-9]*.*", force_show_progress_bar=True)
m.detector.add_step_total_progress_rule("^\[[0-9]*\/\([0-9]*\).*", force_show_progress_bar=True)

m.add_action(["make", "-d", "configure"], can_be_restarted=True, on_success="next_command")
m.add_action(["make", "-d", "build"], can_be_restarted=True, on_success="next_command")
m.add_action(["make", "-d", "test"], can_be_restarted=True)

m.run(run_output="harvest", instance="make", start_instance_if_needed=True, follow=True)
```

Later part can easily be made customizable using vanilla python:

```python3
for i in range(1,len(argv)):
  m.add_restartable_command(["make", "-d", argv[i)], can_be_restarted=True, on_success="next_command")
m.run(run_output="harvest", instance="make", start_instance_if_needed=True, follow=True)
```

**Comment by xavierog** [comment]
> I tried those, but I have troubles understanding if scroll stops, or if scripts takes time. I will use --scroll-on-activity for now anyway so I will figure this later.

They are complementary features: `--scroll-on-activity` provides the automatic scrolling you expect. But sometimes, you may want to read a part of the output while your job is still running. But how can you read step number 5 while Moulti insists on scrolling to display step number 78? The "lock scroll" action enables you to scroll by yourself without being disturbed by the programmatic scrolling.

**Comment by xavierog** [comment]
> Reading all of this, I wonder if a python API might be easier to implement (and more powerful) than JSON/YAML/TMOL.

It is planned to implement a "Python User API", i.e. a Python API exposed to users and that would adhere to semver. It will definitely be more powerful (e.g. it should integrate a class that keeps track of all created steps, and it should not need as many fork-exec-connect-send-receive-close-exit operations as a shell script) but this will likely wait until things get more stable (because a stable feature set makes it easier to design a perennial API).

My reflections so far:
- re: the idea of an output format that Moulti would recognize out of the box: it still makes sense but ideally, this idea should be implemented using one of the ideas below, possibly serving as example/demo.
- re: the idea of a framework for imperative-style parsing/translation: one issue is that there are too many languages and I am reluctant to introduce a new Domain-Specific Language. Considering how it quickly and concisely solved your own issue, I am tempted to make bash a priority: adding a few functions to `moulti-functions.bash`, demonstrating their use in the documentation and ending with a solution that is even more concise.
- re: the idea of a declarative-style tool: the idea remains relevant precisely because it is independent of any programming language. It still needs exploration, but I already pinpointed a use case that requires a Finite-State Machine and thus brings complexity: parsing such titles:
```
==========================================================
This is a title only because it is surrounded with = signs
==========================================================
This line should go inside a Moulti step.
This line should go inside a Moulti step.
This line should go inside a Moulti step.
================================================================
This is another title only because it is surrounded with = signs
================================================================
This line should go inside a Moulti step.
This line should go inside a Moulti step.
This line should go inside a Moulti step.
```

By the way, real life use cases are most welcome to feed those reflections. Do you have any other use case in mind?

**Comment by MartyLake** [comment]
> > Reading all of this, I wonder if a python API might be easier to implement (and more powerful) than JSON/YAML/TMOL.
> 
> It is planned to implement a "Python User API", i.e. a Python API exposed to users and that would adhere to semver. 

I find it hard to achieve the perfect API first try. You can also open a beta python user API which adhere to semver by starting at major version `0` for example ?

> It will definitely be more powerful (e.g. it should integrate a class that keeps track of all created steps, and it should not need as many fork-exec-connect-send-receive-close-exit operations as a shell script) 

I find it easier to launch moulti once, when it is done maybe inspect what I need, then close and continue working. This workflow reduces the number of terminals I need to keep open.

Maybe there is another way to work, which is to have one moulti session, and throw at it work to do from multiple context (I use multiple git worktrees sometimes). But then, I would not know how to interpret the result in case of failure: which context did fail?

> but this will likely wait until things get more stable (because a stable feature set makes it easier to design a perennial API).

Sure, that’s reasonable.

> My reflections so far:
> 
>     * re: the idea of an output format that Moulti would recognize out of the box: it still makes sense but ideally, this idea should be implemented using one of the ideas below, possibly serving as example/demo.
> 
>     * re: the idea of a framework for imperative-style parsing/translation: one issue is that there are too many languages and I am reluctant to introduce a new Domain-Specific Language. Considering how it quickly and concisely solved your own issue, I am tempted to make bash a priority: adding a few functions to `moulti-functions.bash`, demonstrating their use in the documentation and ending with a solution that is even more concise.

Definitely! Focusing on real-case solution, maybe it will become more obvious which small details or small options are missing that could make such functions easier to write.
 
>     * re: the idea of a declarative-style tool: the idea remains relevant precisely because it is independent of any programming language. It still needs exploration, but I already pinpointed a use case that requires a Finite-State Machine and thus brings complexity: parsing such titles:

Finite-State Machine could be avoided here by keeping the last 3 lines (instead of the last 1 line) when deciding if this is a new step: `Only = signs` is a rule simple enough to run at each line on the last 3 lines.

> By the way, real life use cases are most welcome to feed those reflections. Do you have any other use case in mind?

The makefile was a real use case. 

Another possible use case is tailing some build job in a distributed CI system: right now, I can do a "while not finished, `curl | tail`", but maybe moulti could be used to have a dashboard, where each section is following an in-progress build for each supported platform.

Another possible use case is podcast downloading: I select podcast episode from another software, then spawn one terminal per podcast episode to download, wait for the downloads, then clean the finished terminals. I could use moulti to have some queue, where I could append download jobs to it. However, it is not really useful, as all downloads eventually complete, and I can just hide the downloading terminals away while multitasking. 

What are the real use case that you imagined ? I mean, apart from Ansible recipes.

**Reply by xavierog** [reply]
> Another possible use case is tailing some build job in a distributed CI system: right now, I can do a "while not finished, curl | tail", but maybe moulti could be used to have a dashboard, where each section is following an in-progress build for each supported platform.

Indeed. That should be feasible. Feeding multiple steps in parallel is demonstrated in `examples/moulti-pass-concurrency.bash`.

> Another possible use case is podcast downloading: I select podcast episode from another software, then spawn one terminal per podcast episode to download, wait for the downloads, then clean the finished terminals. I could use moulti to have some queue, where I could append download jobs to it. However, it is not really useful, as all downloads eventually complete, and I can just hide the downloading terminals away while multitasking.

If the downloading order is purely sequential, Moulti should do. However, if there are multiple parallel downloads, Moulti may fall short since it offers only one global progress bar.

> What are the real use case that you imagined ? I mean, apart from Ansible recipes.

My main use case has always been scripts with well-defined but numerous steps and a varying amount of output (from no output at all to `rsync -v`). The script that pushed me to develop Moulti respawns a VM of mine, i.e. it creates a new VM, sets up absolutely everything (entire Ansible playbooks become mere steps), migrates data and handles the network switchover from the old VM to the new one.
I also use Moulti to rebuild a bunch of OCI images on a weekly basis.
In both cases, I am **glad** I can navigate so conveniently through steps, and I like the ability to archive all of it by just pressing `s`.
I may also use it to display my bank accounts in the future.

Overall, Moulti ought to be relevant as long as these conditions are met:
1. there is something to display
2. the entire thing does not fit on screen (this is usually addressed using a pager like `more` or `less`)
3. that thing can somehow be subdivided into well-delimited parts

**Comment by xavierog** [comment]
Toying with bash:
Additions to `moulti-functionsn.bash`: https://github.com/xavierog/moulti/blob/2556c4ed0d91e6adad5ba24dfa2c06f727f65ec7/examples/moulti-functions.bash#L74-L115

Rewriting `moultimake.bash` accordingly:
```bash
#!/usr/bin/env bash

# Instruct the Moulti instance to harvest any output not explicitly assigned to
# a step:
export MOULTI_RUN_OUTPUT=harvest
export MOULTI_INSTANCE="make"
[ "${MOULTI_RUN}" ] || exec moulti run -- "$0" "$@"

source moulti-functions.bash

MOULTI_NEW_STEP_PATTERN='Must remake target `(.*).\.'
function moulti_make_step {
	moulti step add "$2" --title="$5" --bottom-text=' ' --scroll-on-activity=-1 && echo "$2"
}
function moulti_inspect_line {
	[[ "${1}" =~ 'Successfully remade target' ]] && \
		moulti step update "${2}" --classes=success && moulti set --no-progress-bar
	[[ "${1}" =~ error ]] && moulti step update "${2}" --classes=error
	if [[ "${1}" =~ \[([0-9]+)/([0-9]+)\] ]]; then # e.g. "[19/20]"
		moulti set --progress-bar --progress="${BASH_REMATCH[1]}" --progress-target="${BASH_REMATCH[2]}"
	fi
	[[ "$1" =~ ${MOULTI_NEW_STEP_PATTERN} ]] || echo "${line}"
}

# Instance title:
moulti set --title="$*"

# Run the command given on the command-line and process each line of its
# output:
"$@" 2>&1 | moulti_process_lines
```

`moulti_process_lines` now relies on three items that users may opt to override as they see fit:
- `MOULTI_NEW_STEP_PATTERN`: an ERE (extended regex) pattern to spot lines that should trigger the creation of a new step
- `moulti_make_step`: a function tasked with actually creating a new step
- `moulti_inspect_line`: a function that controls what gets passed to Moulti steps

I do not think it covers all possible corner cases, but that seems like a practical approach. What do you think?

**Reply by MartyLake** [reply]
This new update works equally well, and the resulting `moultimake.bash` looks good to me!

**Reply by xavierog** [reply]
The whole thing is currently sitting in branch `devel`, including documentation: https://github.com/xavierog/moulti/blob/3872bf712935ac984e00be556ad3421720e110ef/Documentation.md#dealing-with-existing-scripts -- I hope it is clear enough.

**Reply by MartyLake** [reply]
Documentation is clear, good example with simple use case.
![chefkiss](https://github.com/xavierog/moulti/assets/1246794/2fb31f3f-7034-4353-8e21-3966c59f5328)


Maybe another documentation that would help is a `contrib` folder where various unmaintained scripts could be gathered as real life examples, without the burden of maintaining them all yourself?

**Reply by xavierog** [reply]
> Maybe another documentation that would help is a contrib folder where various unmaintained scripts could be gathered as real life examples, without the burden of maintaining them all yourself?

I tend to think both real and imaginary examples should end up in the `examples` directory, where contributions are of course welcome -- but maybe I should add a contribution policy?
Examples do not bring an actual maintenance burden on the long run, but they do bring an initial portability burden when they are introduced. Otherly put, it is relatively easy to write examples on one's own operating system, but porting to other OS can be touchy. And of course, examples are meant to showcase Moulti's abilities but that falls short when a script crashes on `date: lol who needs milliseconds?`

**Reply by MartyLake** [reply]
Right.

Could an "example demo runner" help in that regard?
Examples could be provided with their bootstrap, run & cleanup scripts, a bit like acceptance tests?

Not sure how to script multiple terminals, or specify expectations.

Other possible ways of failing: `sed -i: lol who needs inplace replaces`, and anything regarding regex.

**Reply by xavierog** [reply]
I see two simple things that would help:
1. a new bash function that takes a list of tools that should be available in `$PATH`, outputs the list of missing tools through a Moulti step and exits; that way, one can specify dependencies like `toilet`, `ping` or `bat` in a single line:
    ```bash
    moulti_check_requirements 'toilet' 'ping' 'batcat|bat' 'ret:0:date --iso'
    ```
2. a check list to determine whether a given script is fit for inclusion. 

Example of such a check list:
1. runs on Linux and macOS (i.e. a script not running on *BSD remains acceptable)
2. clean enough according to `shellcheck`
3. leverages `moulti_check_requirements` to check for dependencies

**Reply by xavierog** [reply]
I have started working on `moulti_check_requirements` in the `devel` branch.

**Reply by MartyLake** [reply]
How do you use some branch with python? I found that command `pip3 install --force-reinstall git+https://github.com/xavierog/moulti@devel`, but maybe there is a better idiomatic way to do so ?

Also, for some reason, moulti-functions is not distributed by pip: 
```
$ pip3 show -f moulti | grep moulti-functions
```
returns nothing.

**Reply by xavierog** [reply]
> How do you use some branch with python? I found that command `pip3 install --force-reinstall git+https://github.com/xavierog/moulti@devel`, but maybe there is a better idiomatic way to do so ?

Yes, that seems correct assuming you stick to pip (as opposed to pipx). Of course, it is not exactly relevant right now, because:
1. pip does not ship `moulti-functions.bash` (yet? see details below)
2. there is currently no Python changes in the `devel`  branch (and that's a first for a project that is 97.8% Python)

> Also, for some reason, moulti-functions is not distributed by pip:

You are right; pip is primarily meant to ship Python applications and so far I have not looked into shipping `moulti-functions.bash` through pip.
At some point, pip generates "entry points" (`moulti`, `moulti-askpass`) in a directory that may (or may not) end up in $PATH, thus making `moulti` available in $PATH. Technically, if I could add `moulti-functions.bash` as some kind of static file that is stored directly in that directory, then we would get a neat situation where `moulti-functions.bash` is available in $PATH and thus usable without having to compute its path.

**Reply by xavierog** [reply]
> Technically, if I could add `moulti-functions.bash` as some kind of static file that is stored directly in that directory

Ok, I have found how to do that. The only downside is that `moulti-functions.bash` is made executable. This will pollute shell completion: `moulti<tab>` will yield moulti, moulti-askpass and moulti-functions.bash. Not a big deal though.

**Reply by xavierog** [reply]
The change was made in the `devel` branch. The pip3 command you mentioned earlier should now make `moulti-functions.bash`  available in $PATH. Therefore, scripts should simply need to state `source moulti-functions.bash` (stating an absolute path remains possible though).

**Reply by MartyLake** [reply]
> The change was made in the `devel` branch. The pip3 command you mentioned earlier should now make `moulti-functions.bash` available in $PATH. Therefore, scripts should simply need to state `source moulti-functions.bash` (stating an absolute path remains possible though).

I can confirm now that `source moulti-functions.bash` works without specifying a path \o/

One step less to install moulti and get ready to try examples. Also, makes the bash fonctions discoverable because it can be <tab> completed.

**Reply by xavierog** [reply]
And this is now part of [Moulti 1.9.0](https://github.com/xavierog/moulti/releases/tag/v1.9.0)!

**Reply by xavierog** [reply]
> I see two simple things that would help:
> 
>     1. a new bash function [...]
>     2. a check list to determine whether a given script is fit for inclusion.

Done in  the devel branch: https://github.com/xavierog/moulti/tree/devel/examples

**Comment by MartyLake** [comment]
Would it be possible to collapsed finished green steps automatically ? that would reduce clutter even more.

**Reply by xavierog** [reply]
Sure. In `moulti_inspect_line`, replace:
```bash
moulti step update "${2}" --classes=success && moulti set --no-progress-bar
```

with:
```bash
moulti step update "${2}" --classes=success --collapsed && moulti set --no-progress-bar
```

If you want to insert a 5-second delay between the step turning green and the step being collapsed, use this instead:
```bash
moulti step update "${2}" --classes=success && moulti set --no-progress-bar && { moulti_delayed_collapse "${2}" 5 & }
```

**Comment by MartyLake** [comment]
Hello! Not much to add apart from the fact that I use your script daily :) thanks for creating it!

**Reply by xavierog** [reply]
Glad to read that!
FYI:
- a new feature, similar to `moulti diff`, should arrive soon
- there is a [poll](https://github.com/xavierog/moulti/discussions/6) about future features
